### PR TITLE
feat: Add setup.py for plugins and install dependancies in CI

### PR DESCRIPTION
**Pull request by krish8484**
## Description

Added Alfa module and basic dependancies in `setup.py` which should set the base for the script and also installed `multiqc` in CI along with other dependancies which would be used to run our alfa multiqc script.

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have not reduced the existing code coverage

### CODE REVIEW

**Comment by AngryMaciek** [comment]
This branch is extremely messy, please clean it before merging.

For example:
* This repository will contain distinct plugins for MultiQC. for now we are working on ALFA, but maybe in a few monthts there will be a necessity to add another. Therefore you cannot dump everything in the root directory. Please create a dedicated directory `ALFA` under the root and put everything that is ALFA-related there.
* All of the files: `multiqc-plugins/multiqc_alfa/__init__.py` , `multiqc-plugins/multiqc_alfa/modules/__init__.py`, `multiqc-plugins/multiqc_alfa/modules/__init__.py`, `multiqc-plugins/multiqc_alfa/modules/alfa/__init__.py` and `multiqc-plugins/multiqc_alfa/modules/alfa/alfa.py` are empty files. Why are they here?
* The same applies to: the directory branch: `multiqc-plugins/build/lib/multiqc_alfa/` it is populated with empty files. What is the reasoning here?

---

I am not sure we understood each other. What you have to do is to write the MultiQC cloning into the CI YAML file and then build it from the cloned source, but **including our directory with the plugins (this repo)**. So you will have to adjust the `setup.py` of MultiQC inside the CI as well (as I understand it, since MultiQC is being built with the `setuptools` mechanism of Python, right?). That way the newly built MultiQC will include the custom plugins we are developing here.

Right now, in the CI you have:
```
        run: |
          python -m pip install --upgrade pip
          pip install flake8 multiqc
          python setup.py install
```

So you are installing the "already-built' version from the Python Package Index and later and then building *this* repository. This is not what we want.

Our approach would require much less work from you and less files in the repository in general:
1. Build a MultiQC plugin (python files and metadata)
2. CI: clone the MultiQC
3. CI: adjust the `setup.py` of MultiQC to include the plugin.
4. CI: Build MultiQC
5. CI: Test our plugins inside MultiQC

**Comment by krish8484** [comment]
Hi @AngryMaciek, have you had a chance to look at the example plugin made by the author of MultiQC?
https://github.com/MultiQC/example-plugin

- Note that I had created only one directory for `alfa` which is `multiqc-plugins/multiqc_alfa`, the other directories were auto generated when I ran `setup.py`. This is also the case when you run the `setup.py` of the `example-plugin` whose link I have shared above. Deleting these folders require root access, I think we can delete it for now but they will reappear when we run the `setup.py` file again, so we will have to keep deleting it.
- For all the `__init__.py` files and why are they empty please refer https://stackoverflow.com/questions/448271/what-is-init-py-for#:~:text=__init__.py%20can,also%20set%20the%20appropriate%20content.&text=Hence%20using%20*%20importing%2C%20module%20package,__init__.py%20content. Moreover, these files are also in the example-plugin by the author of MultiQC, he has included a note in the read me which states  
```
Python packages need an __init__.py file in every directory. Here, these are mostly empty (except the one in the my_example folder, which contains a shortcut to make the import statement shorter).
```
Regarding `multiqc-plugins/multiqc_alfa/modules/alfa/alfa.py`, this is where our main code will come in and this is empty just for now.
- About the installation, the method you suggested that is cloning and adjusting the `setup.py` file will definitely include `alfa` but it will also include 100s of other modules mentioned here https://github.com/ewels/MultiQC/tree/master/multiqc/modules and since we are writing a plugin only for alfa and not for these modules, they are not required in this repo, only `alfa` is required.
Also, have a look at https://github.com/MultiQC/example-plugin/blob/master/setup.py, notice that he only included the `example-module`, omitting the others.
Therefore I have installed the "already-built' version from the Python Package Index and later and built this repository, following the steps in `usage` section from https://github.com/MultiQC/example-plugin/blob/master/README.md.

**Comment by AngryMaciek** [comment]
Alright, it seems that I misunderstood how the plugins installation with `setuptools` work, this is great that you found a template example. Good, then we will follow the official scheme.
In that case we can keep all plugins as separate directories under `modules`, right? there is no need to have a `multiqc_alfa` at the beginning.

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -8,8 +8,19 @@ jobs:
     strategy:
       matrix:
         os: [macos-10.15, ubuntu-20.04]
+        python-version: [3.6, 3.7]
```
Comment (AngryMaciek) [inline .github/workflows/build.yml]: No need for the diversification here, we may just use the later version: `3.7`.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -8,8 +8,19 @@ jobs:
...
+        with:
+          python-version: ${{ matrix.python-version }}
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install flake8 multiqc
+          python setup.py install
+      - name: Lint with flake8
+        run: |
+          flake8
```
Comment (AngryMaciek) [inline .github/workflows/build.yml]: Please remove this from here, for the linting we will use a separate Action, take a look at one of my toy repos:  
https://github.com/AngryMaciek/ToySnake/blob/master/.github/workflows/lint.yml
Therefore we will have separate shields for `build` and `flake8` 😉

File: setup.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/env python
+from setuptools import setup, find_packages
+
+version = '0.0.1'
+
+setup(
+    name='multiqc_alfa',
+    version=version,
```
Comment (AngryMaciek) [inline setup.py]: why not `0.0.1` here?

File: setup.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    author='Krish Agarwal',
+    author_email='akrish136@gmail.com',
+    description="MultiQC plugin for the Zavolab \
+        @ University of Basel, Switzerland",
+    long_description=__doc__,
+    keywords='bioinformatics',
+    url='https://github.com/zavolanlab/multiqc-plugins',
+    download_url='',  # after releasing a version
+    license='',  # to add
```
Comment (AngryMaciek) [inline setup.py]: Apache 2.0 is good

**Review by AngryMaciek — COMMENTED** [review]
One more general thing, in addition to all smaller comments:
The tool is written in capitals: https://github.com/biocompibens/ALFA
Please change that in all places/filenames.

File: .github/workflows/build.yml (hunk 2 above)
Comment (AngryMaciek) [reply .github/workflows/build.yml]: Miniconda here is not neccesary, ofc. My point is just to have a separate action with `black` and `flake8` checks.

**Comment by AngryMaciek** [comment]
Could you also update the `README` a little? add maybe a link to the tutorial from MultiQC docs and a link to that example repo of the Author. And a bullet point list with current plugins (one-elemental, for now)

File: setup.py (hunk 3 above)
Comment (krish8484) [reply setup.py]: version over here is a variable, so it is `0.0.1`

Comment (AngryMaciek) [reply setup.py]: Yes, I see this - why is the version a variable and not just encoded here as `version='0.0.1'`?
If you do not use this string anywhere else there is no need for unnecessary variables<?>

Comment (krish8484) [reply setup.py]: Alright, changed it to a string

File: .github/workflows/lint.yml [hunk 5]
```diff
@@ -0,0 +1,22 @@
+name: flake8
```
Comment (AngryMaciek) [inline .github/workflows/lint.yml]: I asked you to create a CI test for `black` and `flake8`, the first one is not here.

Comment (krish8484) [reply .github/workflows/lint.yml]: Added now!

**Pull request merged**